pkg/prom/instance/instance.go

```diff
@@ -306,47 +291,115 @@ func (i *Instance) Run(ctx context.Context) error {
 		)
 	}
 	{
+		sm, err := i.readyScrapeManager.Get()
+		if err != nil {
+			level.Error(i.logger).Log("msg", "failed to get scrape manager")
+			return err
+		}
+
 		// Scrape manager
 		rg.Add(
 			func() error {
-				err := scrapeManager.Run(discovery.SyncCh())
+				err := sm.Run(i.discovery.SyncCh())
 				level.Info(i.logger).Log("msg", "scrape manager stopped")
 				return err
 			},
 			func(err error) {
 				// The scrape manager is closed first to allow us to write staleness
 				// markers without receiving new samples from scraping in the meantime.
 				level.Info(i.logger).Log("msg", "stopping scrape manager...")
-				scrapeManager.Stop()
+				sm.Stop()
 
 				// On a graceful shutdown, write staleness markers. If something went
 				// wrong, then the instance will be relaunched.
 				if err == nil && i.cfg.WriteStaleOnShutdown {
 					level.Info(i.logger).Log("msg", "writing staleness markers...")
-					err := wstore.WriteStalenessMarkers(i.getRemoteWriteTimestamp)
+					err := i.wal.WriteStalenessMarkers(i.getRemoteWriteTimestamp)
 					if err != nil {
 						level.Error(i.logger).Log("msg", "error writing staleness markers", "err", err)
 					}
 				}
 
 				level.Info(i.logger).Log("msg", "closing storage...")
-				if err := storage.Close(); err != nil {
+				if err := i.storage.Close(); err != nil {
 					level.Error(i.logger).Log("msg", "error stopping storage", "err", err)
 				}
 			},
 		)
 	}
 
-	err = rg.Run()
+	err := rg.Run()
 	if err != nil {
 		level.Error(i.logger).Log("msg", "agent instance stopped with error", "err", err)
 	}
 	return err
 }
 
+// initialize sets up the various Prometheus components with their initial
+// settings. initialize will be called each time the Instance is run. Prometheus
+// components cannot be reused after they are stopped so we need to recreate them
+// each run.
+//
+// Components are stored as struct members to allow other methods to access
+// components from different goroutines (as opposed to keeping them completely
+// defined within the Run method where nothing else can access them).
+func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer) error {
```

Comment:
You probably already know this but we need to stop the old components before we start new ones (e.g. `i.wal.Close()`).

 If you're planning to do that outside of this function it becomes a bit hard to follow the whole logic.

If you're planning to do it here, then the name of the function might be a bit deceiving (?)

What I'm trying to say is that depending on what you're thinking to do next there might be work to do here. I'll leave that your judgement.

Reply:
Right, the components are stopped at the end of Run. The initialize method was pulled out to make it easier to grab the mutex. Would more comments or better naming be ideal here?

Reply:
Ah! I put it together now, this is only for creating the new components but not starting/stopping them as that is still done within `Run` via the run group.

Why do you need a lock then? And if you do, shouldn't you be holding it for the whole duration until right before you call `rg.Run()`?

Reply:
We need a lock because another goroutine might [request the active targets](https://github.com/rfratto/agent/blob/032826b89b1aea62a2f10f346c630f98fbd3b37f/pkg/prom/instance/instance.go#L395) at any time (i.e., the uses calls the new `/api/agent/v1/targets` API).

It doesn't matter if the components aren't run yet, we just need to prevent a concurrent read/write to the pointers.

Reply:
🤦 sorry, it all makes sense now - I read this all wrong.